examples/step-89/step-89.cc

```diff
@@ -728,7 +733,9 @@ namespace Step89
           velocity_m.gather_evaluate(src, EvaluationFlags::values);
           velocity_p.gather_evaluate(src, EvaluationFlags::values);
 
-          material_handler->reinit_face(face);
+          material_handler->reinit_face(face); // PM: currently this is not thread-safe. We might need
+          // to split up material_handler: 1) vectors of
+          // materials vs. 2) current state of each tread.
```

Comment:
I didn't think about this. I will have a closer look

Reply:
If I understand it correctly, it would be safe if we have e.g. the following members:
```
Threads::ThreadLocalStorage<FEEvaluation> phi;
Threads::ThreadLocalStorage<scalar> density;
```
we could then update the material with:
```
density.get() = phi.get().read_cell_data(rho);
```
and get it with
```
density.get()
```
?

The `MaterialHandler` is constructed with one thread. Once other threads join, the `ThreadLocalStorage` members are extended correctly?

Similarly, we could make `FERemoteEval` thread safe by storing `ThreadLocalStorage<int>current_cell_batch_index`?